### ISSUE TITLE
Render: huge performance regression between 2.24.2 and 2.26.0

### ISSUE THREAD

**Issue by crudelios** [libsdl-org/SDL#6581]
Hi,

As the title suggests, I'm experiencing a very large performance loss between the current and last release.

Version 2.25.1 also has the performance regression.

For comparison, drawing time went from 2ms to 16ms.

I tried to figure out which commit caused the performance loss but I couldn't.

For reference, I tested on the x64 builds of Windows, both using the default (direct3d) renderer and OpenGL too. I made sure  render batching was enabled and also tried disabling vsync but to no avail.

The game works by getting all the textures from a large (3200x3200) atlas texture and drawing them using `SDL_RenderCopy` functions to a target texture. I haven't changed the rendering code at all between SDL builds.

I'm happy to provide more information if needed.

**Comment by 1bsyl** [comment]
`git bisect` is usually the thing to find which commit is the culprit. but you need to install a git version of SDL ...

**Comment by crudelios** [comment]
That is a good idea.  I'll report later.

**Comment by crudelios** [comment]
Result:

`b7fddb4c0ea2a7d4fcf512ca2ddfd82afd13d6a5 is the first bad commit`

So it's apparently caused by b7fddb4c0ea2a7d4fcf512ca2ddfd82afd13d6a5.

**Comment by crudelios** [comment]
Reverting the commit on the HEAD does not resolve the issue.

Also, removing `SDL_INIT_JOYSTICK` from `SDL_Init` does not fix the issue.

Just in case I messed up with `git bisect`, I'm going to manually test b7fddb4c0ea2a7d4fcf512ca2ddfd82afd13d6a5 against the previous commit.

**Comment by crudelios** [comment]
Sorry, I definitely messed up the bisect. I'll have to try it again.

**Comment by crudelios** [comment]
Okay, this time I got it.

4e375996d3c935e78fe7f83fd12cc529d7b0eb80 is the culprit.

I do call `SDL_GetVersion` a lot during rendering to check for specific rendering support. Removing the call restored performance.

I'm guessing it's due to the `SDL_GetHintBoolean` being constantly called. I'm going to cache the version check, and I am unsure if even is something that should be "fixed" on the SDL side of things.

**Comment by slouken** [comment]
Yes, you only need to call that once at startup. I'm glad you found the issue.

**Comment by icculus** [comment]
Hey, if #6627 gets done, that'll solve this for SDL3, but for SDL2, should we just stick a hack in there to make sure that SDL_GetVersion doesn't have a lot of overhead going forward?

```diff
diff --git a/src/SDL.c b/src/SDL.c
index 6297159e4..cde3434a9 100644
--- a/src/SDL.c
+++ b/src/SDL.c
@@ -519,13 +519,21 @@ SDL_Quit(void)
 void
 SDL_GetVersion(SDL_version * ver)
 {
+    static SDL_bool check_hint = SDL_TRUE;
+    static SDL_bool legacy_version = SDL_FALSE;
+
     if (!ver) {
         return;
     }
 
     SDL_VERSION(ver);
 
-    if (SDL_GetHintBoolean("SDL_LEGACY_VERSION", SDL_FALSE)) {
+    if (check_hint) {
+        check_hint = SDL_FALSE;
+        legacy_version = SDL_GetHintBoolean("SDL_LEGACY_VERSION", SDL_FALSE);
+    }
+
+    if (legacy_version) {
         /* Prior to SDL 2.24.0, the patch version was incremented with every release */
         ver->patch = ver->minor;
         ver->minor = 0;
```

Doesn't have to be a full callback thing, just something to keep this fast.
